tests/integration/ef_tests/runner.py

```diff
@@ -0,0 +1,143 @@
+                evm_address,
+                starknet_address,
+            )
+
+
+async def do_transaction(
+    blocks,
+    get_starknet_address,
+    kakarot: StarknetContract,
+    owner,
+    starknet: StarknetContract,
+):
+    """
+    Execute each transaction in each block using an ef-test BlockchainTest format's `blocks` field.
+
+    See: https://ethereum-tests.readthedocs.io/en/latest/test_filler/blockchain_filler.html#blocks for details.
+    """
+    for block in blocks:
+        transactions = block["transactions"]
+
+        for transaction in transactions:
+            starknet_address = get_starknet_address(int(transaction["sender"], 16))
+
+            await kakarot.eth_send_transaction(
+                to=int(transaction["to"], 16),
+                gas_limit=int(transaction["gasLimit"], 16),
+                gas_price=int(transaction["gasPrice"], 16),
+                value=int(transaction["value"], 16),
+                data=hex_string_to_bytes_array(transaction["data"]),
+            ).execute(caller_address=starknet_address)
+
+            # do we really have to do this manually? (yes)
+            await starknet.state.state.increment_nonce(starknet_address)
+
+
+async def assert_eoa_post_state(
+    address, starknet, starknet_address, expected_post_state
+):
+    actual_nonce = await starknet.state.state.get_nonce_at(starknet_address)
+    expected_nonce = int(expected_post_state["nonce"], 16)
+    assert (
+        actual_nonce == expected_nonce
+    ), f"EOA {address}: {expected_nonce=} is not {actual_nonce=}"
+    assert not expected_post_state[
+        "storage"
+    ], f"EOA {address} should have empty post state, but got {expected_post_state['storage']}"
+
+
+async def assert_contract_post_state(
+    address, contract, starknet, starknet_address, expected_post_state
+):
+    actual_nonce = int((await contract.get_nonce().call()).result.nonce)
+    expected_nonce = int(expected_post_state["nonce"], 16)
+
+    assert (
+        actual_nonce == expected_nonce
+    ), f"Contract {address=}: {expected_nonce=} is not {actual_nonce=}"
+
+    for key, expected_storage in expected_post_state["storage"].items():
+        key = int_to_uint256(int(key, 16))
+        expected_storage = int_to_uint256(int(expected_storage, 16))
+        actual_storage = (await contract.storage(key).call()).result.value
+
+        assert (
+            actual_storage == expected_storage
+        ), f"Contract {address}: expected storage={display_storage(expected_storage)} is not actual_storage={display_storage(actual_storage)}"
+
+
+async def assert_post_state(
+    get_contract_account,
+    get_starknet_address,
+    eth: StarknetContract,
+    kakarot: StarknetContract,
+    starknet: StarknetContract,
+    expected_post_state,
+):
+    for address, post_state in expected_post_state.items():
+        evm_address = int(address, 16)
+        starknet_address = get_starknet_address(evm_address)
+
+        if is_account_eoa(post_state):
```

Comment:
consequently I'd just get the nonce from this `if/else`. I think that the logic is more 1: assert nonce, 2: assert storage than assert if eoa else assert if ca

Reply:
something like

```
for address, post_state in expected_post_state.items():
    evm_address = int(address, 16)
    starknet_address = get_starknet_address(evm_address)
    nonce = ... if is_account_eoa(post_state) else ...
    assert nonce == ...

    if expected_storage:
        contract = get_contract_account(starknet_address)
        for key, expected_storage in expected_post_state["storage"].items():
            key = int_to_uint256(int(key, 16))
            expected_storage = int_to_uint256(int(expected_storage, 16))
            actual_storage = (await contract.storage(key).call()).result.value

            assert (
                actual_storage == expected_storage
            ), f"Contract {address}: expected storage={display_storage(expected_storage)} is not actual_storage={display_storage(actual_storage)}"